services/nuget/nuget-v3-service-family.js

```diff
@@ -184,21 +177,26 @@ function createServiceFamily({
         withFeed,
         feed,
       })
-      const { versions } = await fetch(this, { baseUrl, packageName })
-      let latest = versions.slice(-1).pop()
       const includePrereleases = which === 'vpre'
-      if (!includePrereleases) {
-        const filtered = versions.filter(item => {
-          if (semver.valid(item.version)) {
-            return !semver.prerelease(item.version)
-          }
-          return !item.version.includes('-')
+      let data
+      try {
+        data = await fetch(this, {
+          baseUrl,
+          packageName,
+          includePrereleases,
         })
-        if (filtered.length) {
-          latest = filtered.slice(-1).pop()
+      } catch (e) {
+        if (e instanceof NotFound && !includePrereleases) {
+          data = await fetch(this, {
+            baseUrl,
+            packageName,
+            includePrereleases: true,
+          })
+        } else {
+          throw e
         }
```

Comment:
The downside of this approach is that if a user requests a `/v` badge and there are no stable releases, we throw a `NotFound`. Normally if there are no stable versions, but the user requests a `/v` badge, we just show the latest unstable version. This is the case for other version badges on shields and its also what we've done in the past for nuget/myget. In order to reproduce that behaviour (and avoid making a breaking change), we now need to make 2 API requests. This means in the case where we're requesting `/v` but there aren't any stable versions published, this is slower to get the data. Its probably inside the critical 4 second window, but its definitely noticeably slower.

The other thing we could do here is throw an error in the case where the user requests a `/v` badge and there are no stable releases, but its a breaking change and inconsistent with what we do for other services.

I suspect this problem may be one of the reasons we decided to try and decide ourselves on what is the latest version ourselves in the first place instead of just deferring to the registry's opinion in the first place?

Reply:
> its a breaking change and inconsistent with what we do for other services.

Definitely agreed, though this is already a potentially breaking change anyway isn't it?

My sense is that nuget, and accordingly how we handle nuget-related badges, just seems to be different than the other registries we integrate with. We use different rendering logic/color schemes, there's an undocumented/uncommon logic in version processing on the registry side, etc. 

Perhaps the question ultimately boils down to: should we try to treat nuget like all the other package/artifact type registries that we integrate with to be more consistent with our other package/artifact version badges, or do we handle it differently to account for all of its unique aspects to be more consistent with the upstream provider and the experience/expectations of users.

It seems that our current state was an attempt to satisfy both which is likely why we're having these issues. IMO any change we make is going to cause some amount of a breaking change, so I'd prefer we decide which way we want to go (consistency with nuget or consistency with our other badges), and communicate/publicize any associated changes.

Reply:
> Definitely agreed, though this is already a potentially breaking change anyway isn't it?

I think the behaviour change as proposed just brings us into line with user expectations and is best just viewed as a bugfix. I'd be surprised if anyone is knowingly reliant on our sorting being different than nuget's.

> My sense is that nuget, and accordingly how we handle nuget-related badges, just seems to be different than the other registries we integrate with.

Meh.. kinda. I broadly view it as being like PyPI. We don't try and maintain our own implementation of PEP 440. We just say "hey PyPI - what's the latest (pre-)release of this package?" and report what the registry tells us. If the shields badge said a different thing than the registry, that would be unexpected. We should do the same thing for nuget.

> should we try to treat nuget like all the other package/artifact type registries that we integrate with to be more consistent with our other package/artifact version badges, or do we handle it differently to account for all of its unique aspects to be more consistent with the upstream provider and the experience/expectations of users.

In general, our badges should reflect the norms of the community they serve, but I see the decision on what to do in this case as a practicality relating to performance/render time. The fact that we would have to make 2 HTTP requests in this situation is a consequence of the (clunky) API implementation as opposed to the norms of nuget as a packaging community or a desire for consistency across registries.

This has all got a bit philosophical tbh :smile:

Reply:
> This has all got a bit philosophical tbh 😄

I posed the question, with some context around my thinking, because it directly impacts how I'd respond to your inline comment/question. My intent was not to philosophize anything, so a more explicit attempt below.

>  and inconsistent with what we do for other services

^ My main point is that I don't think that neccessarily matters since there's other aspects of the nuget version badge that are already inconsistent with the other version badges. 

I still think it would be better to make a decision on how to handle the nuget version badges holistically with specific actions (I understand most of these are outside the scope of this PR, but enumerating for completeness)

Option 1 - consistent with the nuget registries
1. Update our rendering logic to automatically strip all build metadata from the version values, just like the nuget registries UI and clients do
    * #3926 
2. Keep the unique 3-color rendering logic for our badge
3. Defer to the registry to make pre-rel vs. rel determination on versions
4. For packages with no release versions, we should feel free to throw an error on the `/v` route if there's concerns that attempting to fallback to pre-release versions could lead to timeouts; we should not feel obligated to mimic this aspect of the other version badges, though it'd be great if we could.

Option 2 - consistent with the other version badges
1. Display the raw version values the registry provides in the API response
2. Drop the custom rendering logic and use the same version badge rendering that all the other badges do
3. Fallback to pre-release versions on the `/v` route if there's no release versions
4. Defer to the registry to make pre-rel vs. rel determination, unless that risks (performance or otherwise) the ability to satisfy the third bullet to fallback to pre-release versions. 
    * If the fallback pattern can't be supported by deferring to the registry, then we should attempt to mimic the logic ourselves to ensure the fallback _can_ be supported

I still feel like the current state of our nuget version badges is selectively consistent, and I'd still prefer we pick one or the other and be fully consistent with one or the other.

Reply:
> For packages with no release versions, we should feel free to throw an error on the `/v` route if there's concerns that attempting to fallback to pre-release versions could lead to timeouts; we should not feel obligated to mimic this aspect of the other version badges, though it'd be great if we could.

I think if we were starting from a position of "we don't have a badge for nuget and we're adding one", I'd be on board with that, but bear in mind that the position we're starting from is that we already have a lot of nuget badges in the wild (nuget is one of the more popular services on shields - we're serving over 100,000 requests a day just on the nuget version badge) and this change would take a lot of badges like this one:

https://img.shields.io/nuget/v/EasyCaching.Core.svg

and replace them with an error. Even if this is the decision we'd make if we were starting from scratch, we do need to take that into account.

Maybe we should just try and patch up our existing implementation if the performance hit is too much?


> Update our rendering logic to automatically strip all build metadata from the version values the registry API gives us

In terms of dropping the metadata portion of the version, is it always seperated by a `+` i.e: `version+metadata`, or is this another one where its complicated/we don't really know? There are some [docs on versioning](https://docs.microsoft.com/en-us/nuget/create-packages/prerelease-packages), but build metadata isn;t mentioned.

Reply:
> we're serving over 100,000 requests a day just on the nuget version badge) and this change would take a lot of badges like this one
> and replace them with an error.

Understood, but especially at scale, we could also be replacing a chunk of those existing/working `/v` badges with timeouts instead, especially if a large portion of those requests are for packages that don't have any release versions. 

While I understand that we wouldn't want to make a change that replaced working badges with an error overnight/without warning, I do want to pose the general question of whether we could make such a change after sufficient communication to users (apologies for the semi-OT question, probably best discussed in Discord)

> Maybe we should just try and patch up our existing implementation if the performance hit is too much?

That's probably the safest approach if fully maintaining the status quo for the forseeable future is required (pre-rel fallback, no timeouts/errors, etc.).

I wish we had a way to do some sort of load testing to get a better feel of how likely the change was to cause timeouts in prod, or even some generic telemetry like the percentage of `/v` version badges that fell back to a pre-release version (which I don't think would violate the no-tracking promise at all).

We could give this approach a go for a trial period to see to what degree, if any, it causes timeouts, and then revert back to patching the existing implementation if the timeouts are pervasive. 

The actual code changes here look fine so if you think the odds of timeouts are low then I can give the PR a 👍

Reply:
I reckon we try this first: #4582
If we are still getting issues raised with that approach, lets revisit it.